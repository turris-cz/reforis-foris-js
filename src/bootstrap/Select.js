/*
 * Copyright (C) 2019-2024 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";

import PropTypes from "prop-types";
import { useUID } from "react-uid";

Select.propTypes = {
    /** Select field Label. */
    label: PropTypes.string.isRequired,
    /** Choices if form of {value : "Label",...}. */
    choices: PropTypes.object.isRequired,
    /** Current value. */
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    /** Help text message. */
    helpText: PropTypes.string,
};

function Select({ label, choices, helpText, ...props }) {
    const uid = useUID();

    const options = Object.keys(choices).map((choice) => (
        <option key={choice} value={choice}>
            {choices[choice]}
        </option>
    ));

    return (
        <div className="mb-3">
            <label className="form-label" htmlFor={uid}>
                {label}
            </label>
            <select className="form-select" id={uid} {...props}>
                {options}
            </select>
            {helpText && (
                <div className="form-text">
                    <small>{helpText}</small>
                </div>
            )}
        </div>
    );
}

export default Select;
