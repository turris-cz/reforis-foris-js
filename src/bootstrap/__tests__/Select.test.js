/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";

import {
    fireEvent,
    getByDisplayValue,
    getByText,
    render,
} from "customTestRender";

import Select from "../Select";

const TEST_CHOICES = {
    1: "one",
    2: "two",
    3: "three",
};

describe("<Select/>", () => {
    let selectContainer;
    const onChangeHandler = jest.fn();
    beforeEach(() => {
        const { container } = render(
            <Select
                label="Test label"
                value="1"
                choices={TEST_CHOICES}
                helpText="Help text"
                onChange={onChangeHandler}
            />
        );
        selectContainer = container;
    });

    it("Test with snapshot.", () => {
        expect(selectContainer).toMatchSnapshot();
    });

    it("Test onChange handling.", () => {
        const select = getByDisplayValue(selectContainer, "one");
        expect(select.value).toBe("1");
        fireEvent.change(select, { target: { value: "2" } });

        const option = getByText(selectContainer, "two");
        expect(onChangeHandler).toBeCalled();

        expect(option.value).toBe("2");
    });
});
