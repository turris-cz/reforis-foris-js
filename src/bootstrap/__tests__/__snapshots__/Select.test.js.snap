// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select/> Test with snapshot. 1`] = `
<div>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
      for="1"
    >
      Test label
    </label>
    <select
      class="form-select"
      id="1"
    >
      <option
        value="1"
      >
        one
      </option>
      <option
        value="2"
      >
        two
      </option>
      <option
        value="3"
      >
        three
      </option>
    </select>
    <div
      class="form-text"
    >
      <small>
        Help text
      </small>
    </div>
  </div>
</div>
`;
